#!/usr/bin/env python3
"""
Executor Health Check ë° ê°„ë‹¨í•œ í…ŒìŠ¤íŠ¸

Executor ì„œë²„ê°€ ì‹¤í–‰ ì¤‘ì¸ì§€ í™•ì¸í•˜ê³  ê°„ë‹¨í•œ í…ŒìŠ¤íŠ¸ë¥¼ ìˆ˜í–‰í•©ë‹ˆë‹¤.
"""

import grpc
import sys

try:
    import executor_pb2
    import executor_pb2_grpc
except ImportError:
    print("âŒ Proto íŒŒì¼ì´ ì»´íŒŒì¼ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤.")
    print("   ì‹¤í–‰: make proto")
    sys.exit(1)


def health_check(host='localhost', port=50051):
    """Health Check í…ŒìŠ¤íŠ¸"""
    print(f"ğŸ” Executor Health Check (ì—°ê²°: {host}:{port})")
    
    try:
        channel = grpc.insecure_channel(f'{host}:{port}')
        stub = executor_pb2_grpc.ExecutorStub(channel)
        
        # Health Check ìš”ì²­
        request = executor_pb2.HealthCheckRequest()
        response = stub.HealthCheck(request, timeout=5)
        
        print("âœ… Health Check ì„±ê³µ!")
        print(f"   - Healthy: {response.healthy}")
        print(f"   - Version: {response.version}")
        print(f"   - Active Matches: {response.active_matches}")
        
        channel.close()
        return True
        
    except grpc.RpcError as e:
        print(f"âŒ gRPC ì˜¤ë¥˜: {e.code()}")
        print(f"   ìƒì„¸: {e.details()}")
        print("\nğŸ’¡ Executor ì„œë²„ê°€ ì‹¤í–‰ ì¤‘ì¸ì§€ í™•ì¸í•˜ì„¸ìš”:")
        print("   python -m executor.server")
        return False
        
    except Exception as e:
        print(f"âŒ ì—°ê²° ì˜¤ë¥˜: {e}")
        print("\nğŸ’¡ Executor ì„œë²„ê°€ ì‹¤í–‰ ì¤‘ì¸ì§€ í™•ì¸í•˜ì„¸ìš”:")
        print("   python -m executor.server")
        return False


def test_match_request(host='localhost', port=50051):
    """ê°„ë‹¨í•œ ë§¤ì¹˜ ìš”ì²­ í…ŒìŠ¤íŠ¸ (ì‹¤ì œ ì‹¤í–‰ ì•ˆí•¨)"""
    print(f"\nğŸ§ª ë§¤ì¹˜ ìš”ì²­ í…ŒìŠ¤íŠ¸ ì¤€ë¹„ (ì—°ê²°: {host}:{port})")
    
    try:
        channel = grpc.insecure_channel(f'{host}:{port}')
        stub = executor_pb2_grpc.ExecutorStub(channel)
        
        # í…ŒìŠ¤íŠ¸ìš© ë§¤ì¹˜ ìš”ì²­ ìƒì„±
        request = executor_pb2.MatchRequest(
            match_id="test_match_001",
            environment="pong",
            agents=[
                executor_pb2.AgentData(
                    agent_id="agent1",
                    docker_image="test-agent:v1",
                    version="1.0",
                    metadata={"name": "Test Agent 1"}
                ),
                executor_pb2.AgentData(
                    agent_id="agent2",
                    docker_image="test-agent:v2",
                    version="1.0",
                    metadata={"name": "Test Agent 2"}
                ),
            ],
            timeout_sec=60,
            record_replay=True
        )
        
        print("ğŸ“¤ ë§¤ì¹˜ ìš”ì²­ êµ¬ì¡°:")
        print(f"   - Match ID: {request.match_id}")
        print(f"   - Environment: {request.environment}")
        print(f"   - Agent 1: {request.agents[0].agent_id} ({request.agents[0].docker_image})")
        print(f"   - Agent 2: {request.agents[1].agent_id} ({request.agents[1].docker_image})")
        print(f"   - Timeout: {request.timeout_sec}ì´ˆ")
        
        print("\nâš ï¸  ì‹¤ì œ ë§¤ì¹˜ ì‹¤í–‰ì€ Agent Docker ì´ë¯¸ì§€ê°€ í•„ìš”í•©ë‹ˆë‹¤.")
        print("   í…ŒìŠ¤íŠ¸ë¥¼ ìœ„í•´ ì‹¤ì œ ìš”ì²­ì€ ë³´ë‚´ì§€ ì•ŠìŠµë‹ˆë‹¤.")
        
        # ì‹¤ì œ ìš”ì²­ì„ ë³´ë‚´ë ¤ë©´ ë‹¤ìŒ ì£¼ì„ì„ í•´ì œí•˜ì„¸ìš”:
        # print("\nğŸ“¡ ë§¤ì¹˜ ì‹¤í–‰ ì¤‘...")
        # response = stub.RunMatch(request, timeout=120)
        # print(f"âœ… ë§¤ì¹˜ ì™„ë£Œ!")
        # print(f"   - Status: {response.status}")
        # print(f"   - Winner: {response.winner_agent_id}")
        
        channel.close()
        return True
        
    except Exception as e:
        print(f"âŒ ì˜¤ë¥˜: {e}")
        return False


def main():
    """ë©”ì¸ ì‹¤í–‰"""
    print("="*60)
    print("  RL Arena Executor í…ŒìŠ¤íŠ¸")
    print("="*60)
    
    # 1. Health Check
    if not health_check():
        sys.exit(1)
    
    # 2. ë§¤ì¹˜ ìš”ì²­ í…ŒìŠ¤íŠ¸
    test_match_request()
    
    print("\n" + "="*60)
    print("âœ… í…ŒìŠ¤íŠ¸ ì™„ë£Œ!")
    print("="*60)
    
    print("\nğŸ’¡ ë‹¤ìŒ ë‹¨ê³„:")
    print("   1. Agent Docker ì´ë¯¸ì§€ ë¹Œë“œ (docs/AGENT_IMAGE_GUIDE.md ì°¸ê³ )")
    print("   2. ì‹¤ì œ ë§¤ì¹˜ ì‹¤í–‰ í…ŒìŠ¤íŠ¸")
    print("   3. Backendì™€ ì—°ë™ (EXECUTOR_INTEGRATION_GUIDE.md ì°¸ê³ )")


if __name__ == "__main__":
    main()
